manual/src/tutorials/effectexamples.etex

```diff
@@ -0,0 +1,666 @@
+message. In the latter case, we resume the computations with the value from the
+other. We consider the other cases to be improperly synchronized programs.
+
+We can now define a second computation that also exchanges two messages:
+
+\begin{caml_example*}{verbatim}
+let comp2 () = perform (Xchg 21) * perform (Xchg 21)
+\end{caml_example*}
+
+Finally, we can run the two computations together:
+
+\begin{caml_example}{verbatim}
+run_both (step comp1) (step comp2)
+\end{caml_example}
+
+
+\section{s:effects-user-threads}{User-level threads}
+
+Let us extend the previous example to arbitrary number of tasks. Many languages
+such as GHC Haskell and Go provide user-level threads as a primitive feature
+provided by the runtime system. With effect handlers, user-level threads and
+their schedulers can be implemented in OCaml itself. Typically, user-level
+threading systems provide "fork" primitive to spawn off a new concurrent task
+and a "yield" primitive to yield control to some other task. Correspondingly,
+we shall declare two effects as follows:
+
+\begin{caml_example*}{verbatim}
+type _ Effect.t += Fork : (unit -> unit) -> unit t
+                 | Yield : unit t
+\end{caml_example*}
+
+"Fork" effect takes a thunk and returns a unit to the performer. "Yield" effect
+is unparameterized, and returns a unit when performed. Let us consider that a
+task performing "Xchg" effect may match with any other task also offering to
+exchange a value.
+
+We shall also define helper functions that simply perform these effects:
+
+\begin{caml_example*}{verbatim}
+let fork f = perform (Fork f)
+let yield () = perform Yield
+let xchg v = perform (Xchg v)
+\end{caml_example*}
+
+A top-level "run" function defines the scheduler:
+
+\begin{caml_example*}{verbatim}
+(* A concurrent round-robin scheduler *)
+let run main =
+  let exchanger = ref None in (* waiting exchanger *)
+  let run_q = Queue.create () in (* scheduler queue *)
+  let enqueue k v =
+    Queue.push (fun _ -> continue k v) run_q
+  in
+  let dequeue () =
+    if Queue.is_empty run_q then () (* done *)
+    else (Queue.pop run_q) ()
+  in
+  let rec spawn f =
+    match_with f () {
+      retc = dequeue;
+      exnc = (fun e ->
+        print_endline (Printexc.to_string e);
+        dequeue ());
+      effc = fun (type a) (eff : a t) ->
+        match eff with
+        | Yield -> Some (fun (k : (a, unit) continuation) ->
+            enqueue k (); dequeue ())
+        | Fork f -> Some (fun (k : (a, unit) continuation) ->
+            enqueue k (); spawn f)
+        | Xchg n -> Some (fun (k : (int, unit) continuation) ->
+            begin match !exchanger with
+            | Some (n', k') ->
+                exchanger := None; enqueue k' n; continue k n'
+            | None -> exchanger := Some (n, k); dequeue ()
+            end)
+        | _ -> None
+    }
+  in
+  spawn main
+\end{caml_example*}
+
+We use a mutable queue "run_q" to hold the scheduler queue. "enqueue" pushes
```

Comment:
Suggestion: replace "pushes" with "inserts", since this is a FIFO queue.
Also, indicate somewhere that a FIFO queue is used because this provides round-robin behavior.